From 2ff0933a261bb5972957e6cda8514ee2153147fa Mon Sep 17 00:00:00 2001
From: Michael McAuliffe <michael.e.mcauliffe@gmail.com>
Date: Fri, 26 Jan 2024 11:55:47 -0800
Subject: [PATCH] Add API declarations

---
 extensions/_pywrapfst.pxd | 38 +++++++++++++++++++-------------------
 setup.py                  |  1 +
 2 files changed, 20 insertions(+), 19 deletions(-)

diff --git extensions/_pywrapfst.pxd extensions/_pywrapfst.pxd
index 771e24c..2d90c69 100644
--- extensions/_pywrapfst.pxd
+++ extensions/_pywrapfst.pxd
@@ -67,7 +67,7 @@ cdef fst.WeightClass _get_WeightClass_or_zero(const string &weight_type,
                                               weight_string) except *
 
 
-cdef class Weight:
+cdef public class Weight [type WeightType, object WeightObject]:
 
   cdef unique_ptr[fst.WeightClass] _weight
 
@@ -103,7 +103,7 @@ ctypedef fst.SymbolTable * SymbolTable_ptr
 ctypedef const fst.SymbolTable * const_SymbolTable_ptr
 
 
-cdef class SymbolTableView:
+cdef public class SymbolTableView [type SymbolTableViewType, object SymbolTableViewObject]:
 
   cdef const fst.SymbolTable *_raw(self)
 
@@ -132,7 +132,7 @@ cdef class SymbolTableView:
   cpdef bytes write_to_string(self)
 
 
-cdef class _EncodeMapperSymbolTableView(SymbolTableView):
+cdef public class _EncodeMapperSymbolTableView(SymbolTableView) [type _EncodeMapperSymbolTableViewType, object _EncodeMapperSymbolTableViewObject]:
 
   # Indicates whether this view is of an input or output SymbolTable
   cdef bool _input_side
@@ -140,7 +140,7 @@ cdef class _EncodeMapperSymbolTableView(SymbolTableView):
   cdef shared_ptr[fst.EncodeMapperClass] _mapper
 
 
-cdef class _FstSymbolTableView(SymbolTableView):
+cdef public class _FstSymbolTableView(SymbolTableView) [type _FstSymbolTableViewType, object _FstSymbolTableViewObject]:
 
   # Indicates whether this view is of an input or output SymbolTable
   cdef bool _input_side
@@ -148,7 +148,7 @@ cdef class _FstSymbolTableView(SymbolTableView):
   cdef shared_ptr[fst.FstClass] _fst
 
 
-cdef class _MutableSymbolTable(SymbolTableView):
+cdef public class _MutableSymbolTable(SymbolTableView) [type _MutableSymbolTableType, object _MutableSymbolTableObject]:
 
   cdef fst.SymbolTable *_mutable_raw(self)
 
@@ -161,7 +161,7 @@ cdef class _MutableSymbolTable(SymbolTableView):
   cpdef void set_name(self, new_name) except *
 
 
-cdef class _MutableFstSymbolTableView(_MutableSymbolTable):
+cdef public class _MutableFstSymbolTableView(_MutableSymbolTable) [type _MutableFstSymbolTableViewType, object _MutableFstSymbolTableViewObject]:
 
   # Indicates whether this view is of an input or output SymbolTable
   cdef bool _input_side
@@ -169,7 +169,7 @@ cdef class _MutableFstSymbolTableView(_MutableSymbolTable):
   cdef shared_ptr[fst.MutableFstClass] _mfst
 
 
-cdef class SymbolTable(_MutableSymbolTable):
+cdef public class SymbolTable(_MutableSymbolTable) [type SymbolTableType, object SymbolTableObject]:
 
   cdef unique_ptr[fst.SymbolTable] _smart_table
 
@@ -192,7 +192,7 @@ cdef SymbolTable _init_SymbolTable(unique_ptr[fst.SymbolTable] table)
 cpdef SymbolTable _read_SymbolTable_from_string(string state)
 
 
-cdef class _SymbolTableIterator:
+cdef public class _SymbolTableIterator [type _SymbolTableIteratorType, object _SymbolTableIteratorObject]:
 
   cdef SymbolTableView _table
   cdef unique_ptr[fst.SymbolTableIterator] _siter
@@ -204,7 +204,7 @@ cdef class _SymbolTableIterator:
 ctypedef fst.EncodeMapperClass * EncodeMapperClass_ptr
 
 
-cdef class EncodeMapper:
+cdef public class EncodeMapper [type EncodeMapperType, object EncodeMapperObject]:
 
   cdef shared_ptr[fst.EncodeMapperClass] _mapper
 
@@ -241,7 +241,7 @@ ctypedef fst.MutableFstClass * MutableFstClass_ptr
 ctypedef fst.VectorFstClass * VectorFstClass_ptr
 
 
-cdef class Fst:
+cdef public class Fst [type FstType, object FstObject]:
 
   cdef shared_ptr[fst.FstClass] _fst
 
@@ -307,7 +307,7 @@ cdef class Fst:
   cpdef bytes write_to_string(self)
 
 
-cdef class MutableFst(Fst):
+cdef public class MutableFst(Fst) [type MutableFstType, object MutableFstObject]:
 
   cdef shared_ptr[fst.MutableFstClass] _mfst
 
@@ -388,7 +388,7 @@ cdef class MutableFst(Fst):
   cdef void _topsort(self)
 
 
-cdef class VectorFst(MutableFst):
+cdef public class VectorFst(MutableFst) [type VectorFstType, object VectorFstObject]:
 
     pass
 
@@ -410,7 +410,7 @@ cpdef Fst _read_Fst_from_string(string state)
 # Iterators.
 
 
-cdef class Arc:
+cdef public class Arc [type ArcType, object ArcObject]:
 
   cdef unique_ptr[fst.ArcClass] _arc
 
@@ -420,7 +420,7 @@ cdef class Arc:
 cdef Arc _init_Arc(const fst.ArcClass &arc)
 
 
-cdef class _ArcIterator:
+cdef public class _ArcIterator [type _ArcIteratorType, object _ArcIteratorObject]:
 
   cdef shared_ptr[fst.FstClass] _fst
   cdef unique_ptr[fst.ArcIteratorClass] _aiter
@@ -442,7 +442,7 @@ cdef class _ArcIterator:
   cdef Arc _value(self)
 
 
-cdef class _MutableArcIterator:
+cdef public class _MutableArcIterator [type _MutableArcIteratorType, object _MutableArcIteratorObject]:
 
   cdef shared_ptr[fst.MutableFstClass] _mfst
   cdef unique_ptr[fst.MutableArcIteratorClass] _aiter
@@ -466,7 +466,7 @@ cdef class _MutableArcIterator:
   cdef Arc _value(self)
 
 
-cdef class _StateIterator:
+cdef public class _StateIterator [type _StateIteratorType, object _StateIteratorObject]:
 
   cdef shared_ptr[fst.FstClass] _fst
   cdef unique_ptr[fst.StateIteratorClass] _siter
@@ -588,7 +588,7 @@ cpdef MutableFst synchronize(Fst ifst)
 # Compiler.
 
 
-cdef class Compiler:
+cdef public class Compiler [type CompilerType, object CompilerObject]:
 
   cdef unique_ptr[stringstream] _sstrm
   cdef string _fst_type
@@ -608,7 +608,7 @@ cdef class Compiler:
 
 # FarReader.
 
-cdef class FarReader:
+cdef public class FarReader [type FarReaderType, object FarReaderObject]:
 
   cdef unique_ptr[fst.FarReaderClass] _reader
 
@@ -633,7 +633,7 @@ cdef class FarReader:
 
 # FarWriter.
 
-cdef class FarWriter:
+cdef public class FarWriter [type FarWriterType, object FarWriterObject]:
 
   cdef unique_ptr[fst.FarWriterClass] _writer
 
diff --git setup.py setup.py
index 5c946d3..71913be 100644
--- setup.py
+++ setup.py
@@ -67,6 +67,7 @@ pywrapfst = Extension(
     extra_compile_args=COMPILE_ARGS,
     libraries=LIBRARIES,
     sources=["extensions/_pywrapfst.pyx"],
+    define_macros=[("CYTHON_EXTERN_C", 'extern "C"')],
 )
 pynini = Extension(
     name="_pynini",
-- 
2.33.0.windows.1

