diff --git a/extensions/stringmap.h b/extensions/stringmap.h
index 6a5e030..2fbaba9 100644
--- a/extensions/stringmap.h
+++ b/extensions/stringmap.h
@@ -182,7 +182,7 @@ bool StringMapCompile(internal::ColumnStringFile *csf, MutableFst<Arc> *fst,
     const auto log_line_compilation_error = [&csf, &line]() {
       LOG(ERROR) << "StringFileCompile: Ill-formed line " << csf->LineNumber()
                  << " in file " << csf->Filename() << ": `"
-                 << fst::StringJoin(line, "\t") << "`";
+                 << fst::StrJoin(line, "\t") << "`";
     };
     switch (line.size()) {
       case 1: {
@@ -227,7 +227,7 @@ bool StringMapCompile(const std::vector<std::vector<std::string>> &lines,
   for (const auto &line : lines) {
     const auto log_line_compilation_error = [&line]() {
       LOG(ERROR) << "StringMapCompile: Ill-formed line: `"
-                 << fst::StringJoin(line, "\t") << "`";
+                 << fst::StrJoin(line, "\t") << "`";
     };
     switch (line.size()) {
       case 1: {
